Sources/NIOPerformanceTester/DatagramChannelBenchmark.swift

```diff
@@ -0,0 +1,142 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+import NIOCore
+import NIOPosix
+
+fileprivate final class DoNothingHandler: ChannelInboundHandler {
+    typealias InboundIn = ByteBuffer
+    typealias OutboundOut = ByteBuffer
+}
+
+class DatagramClientBenchmark {
+    fileprivate final let group: MultiThreadedEventLoopGroup
+    fileprivate final let serverChannel: Channel
+    fileprivate final let localhostPickPort: SocketAddress
+    fileprivate final let clientBootstrap: DatagramBootstrap
+    fileprivate final let clientChannel: Channel
+    fileprivate final let payload: NIOAny
+
+    final let iterations: Int
+
+    fileprivate init(iterations: Int, connect: Bool, envelope: Bool, metadata: Bool) {
+        self.iterations = iterations
+
+        self.group = MultiThreadedEventLoopGroup(numberOfThreads: System.coreCount)
+
+        // server setup
+        self.localhostPickPort = try! SocketAddress.makeAddressResolvingHost("127.0.0.1", port: 0)
+        self.serverChannel = try! DatagramBootstrap(group: self.group)
+            .channelInitializer { $0.pipeline.addHandler(DoNothingHandler()) }
+            .bind(to: localhostPickPort)
+            .wait()
+
+        // client bootstrap setup
+        self.clientBootstrap = DatagramBootstrap(group: self.group)
+            .channelInitializer { $0.pipeline.addHandler(DoNothingHandler()) }
+
+        // client channel setup
+        if connect {
+            self.clientChannel = try! self.clientBootstrap.connect(to: self.serverChannel.localAddress!).wait()
+        } else {
+            self.clientChannel = try! self.clientBootstrap.bind(to: self.localhostPickPort).wait()
+        }
+
+        // payload setup
+        let buffer = self.clientChannel.allocator.buffer(integer: 1, as: UInt8.self)
+        switch (envelope, metadata) {
+        case (true, true):
+            self.payload = NIOAny(AddressedEnvelope<ByteBuffer>(
+                remoteAddress: self.serverChannel.localAddress!,
+                data: buffer,
+                metadata: .init(ecnState: .transportCapableFlag1)
+            ))
+        case (true, false):
+            self.payload = NIOAny(AddressedEnvelope<ByteBuffer>(
+                remoteAddress: self.serverChannel.localAddress!,
+                data: buffer
+            ))
+        case (false, false):
+            self.payload = NIOAny(buffer)
+        case (false, true):
+            preconditionFailure("No API for this")
+        }
+
+        // send one payload to activate the channel
+        try! self.clientChannel.writeAndFlush(payload).wait()
+    }
+
+    func setUp() throws {
+    }
+
+    func tearDown() {
+        try! self.clientChannel.close().wait()
+        try! self.serverChannel.close().wait()
+        try! self.group.syncShutdownGracefully()
+    }
+}
+
+final class DatagramBootstrapCreateBenchmark: DatagramClientBenchmark, Benchmark {
```

Comment:
You're right—throughout this PR—that some of these benchmark classes inherit very little from the base. This one inherits almost nothing. 

The reason I went this way is because I planned to port this code over to the allocation test framework where it would be nice for everything to share some common scaffolding. 

It also makes it clear that everything is set up the same for every benchmark, and then we'll just benchmark a different part of the flow in the critical loop. This is motivated by the discussion in the issue #2187. It would be nice to make it very clear that in all of these tests, we establish a control, and change and measure just one thing.

WDYT?